drivers/adc/adc_stm32.c

```diff
@@ -828,72 +806,89 @@ static void dma_callback(const struct device *dev, void *user_data,
 }
 #endif /* CONFIG_ADC_STM32_DMA */
 
-static int start_read(const struct device *dev,
-		      const struct adc_sequence *sequence)
+static uint8_t get_reg_value(const struct device *dev, uint32_t reg,
+			     uint32_t shift, uint32_t mask)
 {
 	const struct adc_stm32_cfg *config = dev->config;
-	struct adc_stm32_data *data = dev->data;
 	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
-	uint8_t resolution;
-	int err;
 
-	switch (sequence->resolution) {
-#if defined(CONFIG_SOC_SERIES_STM32F1X) || \
-	defined(STM32F3X_ADC_V2_5)
-	case 12:
-		resolution = table_resolution[0];
-		break;
-#elif defined(CONFIG_SOC_SERIES_STM32U5X)
-	case 6:
-		resolution = table_resolution[0];
-		break;
-	case 8:
-		resolution = table_resolution[1];
-		break;
-	case 10:
-		resolution = table_resolution[2];
-		break;
-	case 12:
-		resolution = table_resolution[3];
-		break;
-	case 14:
-		resolution = table_resolution[4];
-		break;
-#elif !defined(CONFIG_SOC_SERIES_STM32H7X)
-	case 6:
-		resolution = table_resolution[0];
-		break;
-	case 8:
-		resolution = table_resolution[1];
-		break;
-	case 10:
-		resolution = table_resolution[2];
-		break;
-	case 12:
-		resolution = table_resolution[3];
-		break;
-#else
-	case 8:
-		resolution = table_resolution[0];
-		break;
-	case 10:
-		resolution = table_resolution[1];
-		break;
-	case 12:
-		resolution = table_resolution[2];
-		break;
-	case 14:
-		resolution = table_resolution[3];
-		break;
-	case 16:
-		resolution = table_resolution[4];
-		break;
-#endif
-	default:
+	uintptr_t addr = (uintptr_t)adc + reg;
+
+	return ((*(volatile uint32_t *)addr >> shift) & mask);
+}
+
+static void set_reg_value(const struct device *dev, uint32_t reg,
+			  uint32_t shift, uint32_t mask, uint32_t value)
+{
+	const struct adc_stm32_cfg *config = dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	uintptr_t addr = (uintptr_t)adc + reg;
+
+	sys_clear_bits(addr, mask << shift);
+	sys_set_bits(addr, value << shift);
```

Comment:
This will do two register writes.  Probably don't want to do that, as writing to the register has side effects.  There are two significant issues with this.

* The compiler can not optimize out the intermediate write.  Instead of generating efficient code to clear and set the field in the a CPU register, then write to the device, it must first clear the field, write this to the device register, then read it back from the device register, then set the bits, then write it again.
* Setting the field to all zeros in the device register might actually do something on the device!

Every decent API for reading or writing register fields I've used or designed includes a function that will set a field to a new value, without multiple writes to the register.  Example from the Linux regmap API:

```c
 regmap_update_bits(struct regmap *map, unsigned int reg, unsigned int mask, unsigned int val);

regmap_update_bits(adc, reg, mask << shift, value << shift);
```

Functions named `*and_or*` is another common pattern for this.  Since the mask is and-ed and the the value is or-ed.

But the Zephyr functions in `sys_bitops.h` don't seem to have it.

Reply:
Agreed, uses MODIFY_REG instead